js/main.ts

```diff
@@ -72,5 +72,7 @@ export default function denoMain() {
 
   const inputFn = argv[1];
   const mod = runtime.resolveModule(inputFn, `${cwd}/`);
-  mod.compileAndRun();
+  assert(mod != null);
+  // TypeScript does not track assert, therefore not null assertion
+  mod!.compileAndRun();
```

Comment:
These null assert checks are redundant when using the bang operator. Suggest removing the assert.

Reply:
@ry @kitsonk FYI, you can avoid the bang if you create a new `assertNotNull` function like so:


```ts
function assertNotNull<T>(val: T | null | undefined, msg = "assert not null"): val is T {
  if (val === null || val === undefined) {
    throw Error(msg);
  }
  return true;
}
```

See this question on stackoverflow for more info: https://stackoverflow.com/a/46700791/266535

Reply:
@styfle that doesn't work the way you think it would work, because TypeScript does not _know_ that the code past that point is narrowed without some block scoping.  For example, if [this is run under `--strictNullChecks`](https://www.typescriptlang.org/play/#src=function%20assertNotNull%3CT%3E(val%3A%20T%20%7C%20null%20%7C%20undefined%2C%20msg%20%3D%20%22assert%20not%20null%22)%3A%20val%20is%20T%20%7B%0D%0A%20%20if%20(val%20%3D%3D%3D%20null%20%7C%7C%20val%20%3D%3D%3D%20undefined)%20%7B%0D%0A%20%20%20%20throw%20Error(msg)%3B%0D%0A%20%20%7D%0D%0A%20%20return%20true%3B%0D%0A%7D%0D%0A%0D%0Adeclare%20const%20foo%3A%20string%20%7C%20undefined%3B%0D%0Adeclare%20function%20bar(val%3A%20string)%3A%20void%3B%0D%0A%0D%0Afoo%3B%20%2F%2F%20string%20%7C%20undefined%0D%0AassertNotNull(foo)%3B%0D%0Afoo%3B%20%2F%2F%20string%20%7C%20undefined%0D%0Abar(foo)%3B%20%2F%2F%20maybe%20undefined%0D%0A):

```ts
function assertNotNull<T>(val: T | null | undefined, msg = "assert not null"): val is T {
  if (val === null || val === undefined) {
    throw Error(msg);
  }
  return true;
}

declare const foo: string | undefined;
declare function bar(val: string): void;

foo; // string | undefined
assertNotNull(foo);
foo; // string | undefined
bar(foo); // maybe undefined
```

Also see Microsoft/TypeScript#8655 which explains that problem is not currently supported by TypeScript.

Reply:
@kitsonk Oh right, you would need an `if` statement for the assert for it to narrow the type as not null.
But it's still safer than using `!` in my opinion 😅